File: Evals/IGES-SDK/tests/spec/test_writer_global.cpp
// Tests for Global Section writer — round-trip.
// Written BEFORE implementation (TDD red phase).

#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers_floating_point.hpp>
#include "writer/global_writer.hpp"
#include "model/global_section.hpp"

using namespace iges;
using Catch::Matchers::WithinRel;

// ─────────────────────────────────────────────────────────────────
// Round-trip: serialize → parse → compare
// ─────────────────────────────────────────────────────────────────

TEST_CASE("Global section round-trip — all 26 fields", "[writer][round-trip]") {
    GlobalSection g;
    g.param_delimiter = ',';
    g.record_delimiter = ';';
    g.product_id_sender = "TestProduct";
    g.file_name = "test.igs";
    g.native_system_id = "IGES-SDK";
    g.preprocessor_version = "v0.1";
    g.integer_bits = 32;
    g.sp_magnitude = 38;
    g.sp_significance = 6;
    g.dp_magnitude = 308;
    g.dp_significance = 15;
    g.product_id_receiver = "Receiver";
    g.model_space_scale = 1.0;
    g.units = Units::Millimeters;
    g.units_name = "MM";
    g.max_line_weight_grads = 4;
    g.max_line_weight_width = 0.01;
    g.file_timestamp = {2026, 4, 11, 12, 0, 0};
    g.min_resolution = 0.0001;
    g.max_coordinate = 1000.0;
    g.author = "John";
    g.organization = "Company";
    g.spec_version = SpecVersion::V5_3;
    g.drafting_std = DraftingStandard::None;
    g.model_timestamp = Timestamp{2026, 4, 11, 12, 0, 0};
    g.app_protocol = "";

    std::string serialized = write_global_section(g);

    // Parse it back
    auto r = parse_global_section(serialized);
    REQUIRE(r.has_value());
    auto& g2 = r.value();

    CHECK(g2.param_delimiter == ',');
    CHECK(g2.record_delimiter == ';');
    CHECK(g2.product_id_sender == "TestProduct");
    CHECK(g2.file_name == "test.igs");
    CHECK(g2.native_system_id == "IGES-SDK");
    CHECK(g2.preprocessor_version == "v0.1");
    CHECK(g2.integer_bits == 32);
    CHECK(g2.sp_magnitude == 38);
    CHECK(g2.sp_significance == 6);
    CHECK(g2.dp_magnitude == 308);
    CHECK(g2.dp_significance == 15);
    CHECK(g2.product_id_receiver == "Receiver");
    CHECK_THAT(g2.model_space_scale, WithinRel(1.0));
    CHECK(g2.units == Units::Millimeters);
    CHECK(g2.units_name == "MM");
    CHECK(g2.max_line_weight_grads == 4);
    CHECK_THAT(g2.max_line_weight_width, WithinRel(0.01));
    CHECK(g2.file_timestamp.year == 2026);
    CHECK_THAT(g2.min_resolution, WithinRel(0.0001));
    CHECK_THAT(g2.max_coordinate, WithinRel(1000.0));
    CHECK(g2.author == "John");
    CHECK(g2.organization == "Company");
    CHECK(g2.spec_version == SpecVersion::V5_3);
    CHECK(g2.drafting_std == DraftingStandard::None);
    REQUIRE(g2.model_timestamp.has_value());
    CHECK(g2.model_timestamp->year == 2026);
}

TEST_CASE("Global section round-trip — minimal with defaults", "[writer][round-trip]") {
    GlobalSection g;
    g.product_id_sender = "Min";
    g.file_name = "min.igs";
    g.native_system_id = "SDK";
    g.preprocessor_version = "v1";
    g.integer_bits = 32;
    g.sp_magnitude = 38;
    g.sp_significance = 6;
    g.dp_magnitude = 308;
    g.dp_significance = 15;
    g.max_line_weight_width = 0.01;
    g.file_timestamp = {2026, 1, 1, 0, 0, 0};
    g.min_resolution = 0.001;

    std::string serialized = write_global_section(g);
    auto r = parse_global_section(serialized);
    REQUIRE(r.has_value());
    CHECK(r.value().product_id_sender == "Min");
    CHECK(r.value().units == Units::Inches);  // default
    CHECK_THAT(r.value().model_space_scale, WithinRel(1.0));  // default
}

File: Evals/IGES/tests/test_writer_global.py
"""Writer-focused tests for Global-section serialization."""
# pyright: reportUnknownMemberType=none
# pyright: reportUnknownVariableType=none
# pyright: reportUnknownArgumentType=none
from __future__ import annotations

from collections.abc import Sequence
from pathlib import Path

from iges_support import parse_iges_to_json, wrap_entities, write_iges_from_json


def test_write_and_parse_preserve_all_26_global_fields(
    submission_command: Sequence[str], tmp_path: Path
) -> None:
    doc = wrap_entities([])
    global_section = doc["global"]
    global_section.update({
        "param_delimiter": "|",
        "record_delimiter": "#",
        "product_id_sender": "WriterProduct",
        "file_name": "writer-full.igs",
        "native_system_id": "WRITER-SDK",
        "preprocessor_version": "v9.9",
        "integer_bits": 64,
        "sp_magnitude": 37,
        "sp_significance": 6,
        "dp_magnitude": 307,
        "dp_significance": 15,
        "product_id_receiver": "Receiver",
        "model_space_scale": 0.125,
        "units": "millimeters",
        "units_name": "MM",
        "max_line_weight_grads": 4,
        "max_line_weight_width": 0.02,
        "file_timestamp": {
            "year": 2026,
            "month": 4,
            "day": 16,
            "hour": 13,
            "minute": 45,
            "second": 30,
        },
        "min_resolution": 1.0e-6,
        "max_coordinate": 2500.0,
        "author": "Jane Doe",
        "organization": "ACME CAD",
        "spec_version": "v5_3",
        "drafting_std": "ansi",
        "model_timestamp": {
            "year": 2026,
            "month": 4,
            "day": 15,
            "hour": 8,
            "minute": 0,
            "second": 5,
        },
        "app_protocol": "AP242",
    })

    iges_path = write_iges_from_json(submission_command, doc, tmp_path, name="global-full")
    parsed = parse_iges_to_json(submission_command, iges_path, tmp_path, name="global-full")

    assert parsed["entities"] == []
    assert parsed["global"] == global_section


def test_write_empty_document_roundtrips_default_global_values(
    submission_command: Sequence[str], tmp_path: Path
) -> None:
    doc = wrap_entities([])
    iges_path = write_iges_from_json(submission_command, doc, tmp_path, name="global-defaults")
    parsed = parse_iges_to_json(submission_command, iges_path, tmp_path, name="global-defaults")

    global_section = parsed["global"]
    assert isinstance(global_section, dict)
    assert global_section["product_id_sender"] == "TEST"
    assert global_section["product_id_receiver"] == "TEST"
    assert global_section["units"] == "inches"
    assert global_section["model_space_scale"] == 1.0
    assert global_section["model_timestamp"] is None
    assert global_section["app_protocol"] == ""
